### PR TITLE
TOOL-1134 Added APIKey and Issuer ID as input to enable JWT authentic…

### PR DESCRIPTION
**Pull request by richard-bogdan-bitrise**
…ation for uploading.

### CODE REVIEW

**Comment by richard-bogdan-bitrise** [comment]
@trapacska: I suggest minor version upgrade for the release, as new inputs (and functionality) were added. Do you agree?

File: main.go [hunk 1]
```diff
@@ -21,10 +26,14 @@ type Config struct {
 	ItunesConnectUser string          `env:"itunescon_user"`
 	Password          stepconf.Secret `env:"password"`
 	AppPassword       stepconf.Secret `env:"app_password"`
-	APIKey            string          `env:"api_key"`
+	APIKeyPath        string          `env:"api_key_path"`
 	APIIssuer         string          `env:"api_issuer"`
 }
 
+var keyPaths = []string{"./private_keys", "~/private_keys", "~/.appstoreconnect/private_keys"}
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: Added

File: main.go [hunk 2]
```diff
@@ -21,10 +26,14 @@ type Config struct {
...
+
+const keyFormat = "AuthKey_(.+)\\.p8"
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

File: main.go [hunk 3]
```diff
@@ -74,29 +81,237 @@ func main() {
...
 	log.Infof("Uploading - %s ...", fileName)
 	log.Printf("$ %s", commandStr)
+	return cmd, nil
+}
 
-	if err := cmd.Run(); err != nil {
-		failf("Uploading IPA failed: %s", err)
+// getAuthOptions provides the command options for authentication. Either a user and password pair or API key and issuer
+// ID should be set in the config.
+func getAuthOptions(cfg Config) ([]string, string, error) {
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: They are in order already, I updated only 1 methods order only.

File: main.go [hunk 4]
```diff
@@ -74,29 +81,237 @@ func main() {
...
+	passOptions, redact, passErr := getUserAndPasswordOptions(cfg)
+	if passErr != nil {
+		apiOptions, apiErr := getAPIKeyAndIssuerOptions(cfg)
+		if apiErr != nil {
+			return []string{}, "", fmt.Errorf("neither user and password or API key and issuer ID is provided correctly, please define one of the pairs in the config.\nIssue with user and password: %s\nIssue with API key and issuer ID: %s", passErr, apiErr)
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: Moved the related code next to validation.

File: main.go [hunk 5]
```diff
@@ -74,29 +81,237 @@ func main() {
...
+	}
+	password := getPasswordFromConfig(cfg)
+	if password == "" {
+		return []string{}, "", fmt.Errorf("neither password or app password is configured, please define it the config")
+	}
+	return []string{"-u", cfg.ItunesConnectUser, "-p", password}, password, nil
+}
+
+// getPasswordFromConfig gets the password from the config.
+func getPasswordFromConfig(cfg Config) string {
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: Not sure what you want and why, please explain.

File: main.go [hunk 6]
```diff
@@ -74,29 +81,237 @@ func main() {
...
+		}
+		return apiOptions, "", nil
 	}
+	return passOptions, redact, nil
+}
 
-	fmt.Println()
-	log.Donef("IPA uploaded")
+// getUserAndPasswordOptions provides the user and password pair for authentication.
+func getUserAndPasswordOptions(cfg Config) ([]string, string, error) {
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: Actually writing small functions is a good coding principle in modern coding. Please have a look at:
[here](https://medium.com/oril/10-tips-to-keep-your-code-more-clean-2fa9aafea1cf)
or
[here](https://martinfowler.com/bliki/FunctionLength.html)
or
[here](https://refactoring.guru/smells/long-method)

File: main.go [hunk 7]
```diff
@@ -74,29 +81,237 @@ func main() {
...
+	}
+	if err := download(url, dstPath); err != nil {
+		return "", fmt.Errorf("failed to download keystore, error: %s", err)
+	}
+	return getAPIKeyFromPath(dstPath)
+}
+
+// checkKeyFormat checks if the given key's name matches the required format by altool.
+func checkKeyFormat(keyName string) bool {
+	if _, err := getAPIKeyFromFileName(keyName); err != nil {
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: This is easier to read:
```
if checkKeyFormat(keyName) && checkKeyExistsAtDirs(keyName, keyDirs)
```
than
```
if _, err := getAPIKeyFromFileName(keyName); err != nil && checkKeyExistsAtDirs(keyName, keyDirs)
```

File: main.go [hunk 8]
```diff
@@ -74,29 +81,237 @@ func main() {
...
+// fileExists checks if the given file exists (and not a directory) or not.
+func fileExists(filename string) bool {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return !info.IsDir()
+}
+
+// getDstPath gets the destination path for the key.
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

File: main.go [hunk 9]
```diff
@@ -74,29 +81,237 @@ func main() {
...
+func getDstPath(keyPath string) (string, error) {
+	dstName, err := getDstName(path.Base(keyPath))
+	if err != nil {
+		return "", fmt.Errorf("could not get destination path. Error: %s", err)
+	}
+	return path.Join(keyPaths[0], dstName), nil
+}
+
+// getDstName gets the name for the key.
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

File: step.yml [hunk 10]
```diff
@@ -64,3 +71,28 @@ inputs:
     description: |-
       Path to your PKG file to be deployed.
       **NOTE:** This input or `IPA path` is required.
+- api_key_path: ""
+  opts:
+    title: "API Key"
+    description: |-
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply step.yml]: Updated.

File: main.go [hunk 11]
```diff
@@ -74,29 +81,237 @@ func main() {
...
+// getUserAndPasswordOptions provides the user and password pair for authentication.
+func getUserAndPasswordOptions(cfg Config) ([]string, string, error) {
+	if cfg.ItunesConnectUser == "" {
+		return []string{}, "", fmt.Errorf("iTunes Connect user is not configured, please define it in the config")
+	}
+	password := getPasswordFromConfig(cfg)
+	if password == "" {
+		return []string{}, "", fmt.Errorf("neither password or app password is configured, please define it the config")
+	}
+	return []string{"-u", cfg.ItunesConnectUser, "-p", password}, password, nil
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: Sometimes you have to return multiple values or introduce a type for it. I didn't introduce new type, as 3 values are not that much.

File: main.go [hunk 12]
```diff
@@ -74,29 +81,237 @@ func main() {
...
+func getPasswordFromConfig(cfg Config) string {
+	password := string(cfg.Password)
+	if string(cfg.AppPassword) != "" {
+		password = string(cfg.AppPassword)
+	}
+	return password
+}
+
+// getAPIKeyAndIssuerOptions provides the API key and issuer ID pair for authentication.
+func getAPIKeyAndIssuerOptions(cfg Config) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (richard-bogdan-bitrise) [reply main.go]: Code updated.

**Comment by richard-bogdan-bitrise** [comment]
Updated code, please check.

File: main.go [hunk 13]
```diff
@@ -17,24 +23,24 @@ import (
...
 	PkgPath           string          `env:"pkg_path"`
-	ItunesConnectUser string          `env:"itunescon_user,required"`
+	ItunesConnectUser string          `env:"itunescon_user"`
 	Password          stepconf.Secret `env:"password"`
 	AppPassword       stepconf.Secret `env:"app_password"`
+	APIKeyPath        string          `env:"api_key_path"`
+	APIIssuer         string          `env:"api_issuer"`
 }
 
+var keyPaths = []string{"./private_keys", "~/private_keys", "~/.private_keys", "~/.appstoreconnect/private_keys"}
```
Comment (godrei) [inline main.go]: pls comment a reference here

**Pull request closed without merging**